keras_retinanet/models/retinanet.py

```diff
@@ -124,42 +124,68 @@ def default_regression_model(num_values, num_anchors, pyramid_feature_size=256,
+        pyramid_levels: pyramid levels in use.
         feature_size : The feature size to use for the resulting feature levels.
 
     Returns
-        A list of feature levels [P3, P4, P5, P6, P7].
+        A list of feature levels. P3, P4, P5, P6 are always included. P2 and P7 included if in use.
     """
+
+    output_layers = {}
+
     # upsample C5 to get P5 from the FPN paper
     P5           = keras.layers.Conv2D(feature_size, kernel_size=1, strides=1, padding='same', name='C5_reduced')(C5)
     P5_upsampled = layers.UpsampleLike(name='P5_upsampled')([P5, C4])
     P5           = keras.layers.Conv2D(feature_size, kernel_size=3, strides=1, padding='same', name='P5')(P5)
+    output_layers["P5"] = P5
 
     # add P5 elementwise to C4
     P4           = keras.layers.Conv2D(feature_size, kernel_size=1, strides=1, padding='same', name='C4_reduced')(C4)
     P4           = keras.layers.Add(name='P4_merged')([P5_upsampled, P4])
     P4_upsampled = layers.UpsampleLike(name='P4_upsampled')([P4, C3])
     P4           = keras.layers.Conv2D(feature_size, kernel_size=3, strides=1, padding='same', name='P4')(P4)
+    output_layers["P4"] = P4
 
     # add P4 elementwise to C3
     P3 = keras.layers.Conv2D(feature_size, kernel_size=1, strides=1, padding='same', name='C3_reduced')(C3)
     P3 = keras.layers.Add(name='P3_merged')([P4_upsampled, P3])
+    if C2 is not None and 2 in pyramid_levels:
+        P3_upsampled = layers.UpsampleLike(name='P3_upsampled')([P3, C2])
     P3 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=1, padding='same', name='P3')(P3)
+    output_layers["P3"] = P3
+
+    if C2 is not None and 2 in pyramid_levels:
+        P2 = keras.layers.Conv2D(feature_size, kernel_size=1, strides=1, padding='same', name='C2_reduced')(C2)
+        P2 = keras.layers.Add(name='P2_merged')([P3_upsampled, P2])
+        P2 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=1, padding='same', name='P2')(P2)
+        output_layers["P2"] = P2
+
+
 
     # "P6 is obtained via a 3x3 stride-2 conv on C5"
-    P6 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=2, padding='same', name='P6')(C5)
+    if 6 in pyramid_levels:
+        P6 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=2, padding='same', name='P6')(C5)
+        output_layers["P6"] = P6
+
 
     # "P7 is computed by applying ReLU followed by a 3x3 stride-2 conv on P6"
-    P7 = keras.layers.Activation('relu', name='C6_relu')(P6)
-    P7 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=2, padding='same', name='P7')(P7)
+    if 7 in pyramid_levels:
+        if 6 not in pyramid_levels:
+            raise ValueError("P6 is required to use P7")
+        P7 = keras.layers.Activation('relu', name='C6_relu')(P6)
+        P7 = keras.layers.Conv2D(feature_size, kernel_size=3, strides=2, padding='same', name='P7')(P7)
+        output_layers["P7"] = P7
+
+    return [output_layers['P'+str(p)] for p in pyramid_levels]
```

Comment:
Prefer to use `'P{}'.format(p)` for p in pyramid_levels`

Reply:
fixed